### ISSUE TITLE
gogoproto  custom type supported in Node js

### ISSUE THREAD

**Issue by vlasfama** [gogo/protobuf#540]
I am using the gogo/proto in my project.Backend code is written in go lang and i am using grpc for remote procedure call its working fine, But i want to use the same proto file for Node js.
But its throws the error "gogoproto.customtype" not supported. Is it possible to use "gogoproto.customtype" in Node Js. Please suggest me how i can use gogo/proto for node js

**Comment by jmarais** [comment]
>  Is it possible to use "gogoproto.customtype" in Node Js.

Unfortunately not. The ```customtype``` extension is only available to golang code. By using the ```customtype``` option gogo/protobuf will only generate the appropriate golang code in order to handle your specified custom type. 
By using the custom type you will need to satisfy these methods:
https://github.com/gogo/protobuf/blob/master/custom_types.md#custom-type-method-signatures
So if you want to unmarshal the message in NodeJS you would have to write your own unmarshal function in javascript, otherwise it will just be a byte slice.

If it can work, you can try to use the ```casttype``` option. On the golang side this will allow you to cast the original protocol buffer type to your own golang type. When you marshal your custom type it will marshal to a protocol buffer type. This means when you then unmarshal it in NodeJS, that field would just be to oiginal protocolbuffer type and will be understood by the js side.

> But its throws the error "gogoproto.customtype" not supported.

Where are you getting this error?

> Please suggest me how i can use gogo/proto for node js

@awalterschulze This is one thing we didn't discuss (js). I think the current gopherjs support is just to build the  ```github.com/gogo/protobuf/protoc-gen-gogo``` package, which I would guess just lets you generate golang code from js? Is my assumption correct?

**Comment by awalterschulze** [comment]
Yes @jmarais the only reason gogoprotobuf compiles to javascript with gopherjs is so that developers can build bigger projects in go, that depends on gogoprotobuf and can also be compiled to javascript.
gogoprotobuf is not meant to generate nice usable code for javascript only for go.
But this only concerns compiling go to javascript, which is not what I think @nagarajmanjunath is referring to.

Ideally the same proto file should be usable to generate code for Go using gogoprotobuf, Nodejs using some nodejs protobuf generator and java using protoc --java_out.  Unfortunately this does not always work, because some code generators only support proto3 and gogoprotobuf relies on extensions, like customtype, nullable, etc on the descriptor.proto.  Extensions are only supported in proto2.  @nagarajmanjunath which code generator are you using for Nodejs and does it support proto2?

**Comment by jmarais** [comment]
> Yes @jmarais the only reason gogoprotobuf compiles to javascript with gopherjs is so that developers can build bigger projects in go, that depends on gogoprotobuf and can also be compiled to javascript.
gogoprotobuf is not meant to generate nice usable code for javascript only for go.
But this only concerns compiling go to javascript, which is not what I think @nagarajmanjunath is referring to.

Thanks for clearing that up.

> because some code generators only support proto3
Ah yes. Thanks for the reminder.

**Comment by awalterschulze** [comment]
I think might be a "duplicate" of https://github.com/gogo/protobuf/issues/324
But @nagarajmanjunath would need to confirm.
In that case, I would recommend looking at #324 , since @yaronsumel had a work around.

**Comment by nagarajmanjunath** [comment]
@awalterschulze,@jmarais thank you for your response, below is my code which  i am trying to impliment. 
  I have proto file with structure has show below. 
  
    syntax = 'proto3';

    package proto3;

    import "github.com/gogo/protobuf/gogoproto/gogo.proto";
    import "google/protobuf/timestamp.proto";

    option (gogoproto.marshaler_all) = false;
    option (gogoproto.unmarshaler_all) = false;
    option (gogoproto.sizer_all) = true;
    option (gogoproto.goproto_registration) = true;
    option (gogoproto.messagename_all) = true;
    option (gogoproto.protosizer_all) =false;

    service BlockChain {
        rpc GetConsensusState(Empty) returns (ConsensusResponse)
     }

    message ConsensusResponse  {
    bytes RoundState =1 [(gogoproto.customtype)=
      "github.com/tendermint/tendermint/consensus/types.RoundStateSimple",(gogoproto.nullable) =
          false];
    repeated ConsensusResponse PeerRoundStates = 2  [(gogoproto.customtype) = 
      "github.com/tendermint/tendermint/consensus/types.PeerRoundState",(gogoproto.nullable) = 
         false];
       }
  
    message Empty {
     }

   cmd  to generate the protobuf, It working fine.
        
     **protoc -I/usr/local/include -I. -I$GOPATH/src -I$GOPATH/src/github.com/grpc-ecosystem/grpc-gateway/third_party/googleapis --go_out=plugins=grpc:./ ./protos/example.proto**

If i want to use the above proto file in node js,  i ll get error  "gogoproto.customtype not supported"
        
          *Index.js*

           const repl = require('repl');
           const grpc = require('grpc')
           const PROTO_PATH = __dirname + './protos/example.proto'
           const example_proto = grpc.load(PROTO_PATH).example
           console.log(example_proto)

Please suggest me how i can use proto file with custom type in node js.

**Comment by jmarais** [comment]
Hi. @nagarajmanjunath 
As @awalterschulze mentioned it might be the same issue as this case [#324](https://github.com/gogo/protobuf/issues/324).
The issue in that case was that the php generator could not use the gogo.proto file because the php generator could only use:
```syntax = 'proto3';```
proto files.

And our [gogo.proto](https://github.com/gogo/protobuf/blob/master/gogoproto/gogo.proto) file is:
```syntax = "proto2";```
Because we use the extensions feature of proto2.

From your example:
```javascript
const grpc = require('grpc')
const example_proto = grpc.load(PROTO_PATH).example
```
It seems like you are using the ```grpc``` package to load the proto file. Do you know if that package can handle:
```syntax = "proto2";```
files?

**Comment by nagarajmanjunath** [comment]
@jmarais  I am using 

       "dependencies": {
          "grpc": "^1.0.1"
         }
  This package does not support 

       syntax = "proto2"

**Comment by jmarais** [comment]
Ah ok. Yeh, you might then have a problem trying to load your proto file containing gogoprotobuf extensions.

Do you maybe want to bring up this issue with that ```grpc``` project?

**Comment by nagarajmanjunath** [comment]
yes, If this issue is fixed and it will helpful for everyone @jmarais

**Comment by awalterschulze** [comment]
Good luck.
You can tell them that you won't need full proto2 support.
You don't need code generation for proto2, but you do need parsing.

Also maybe you want to try `--gogo_out` instead of `--go_out` so that the go code generator is actually gogoprotobuf and not golang/protobuf.

Also maybe there is a chance that you simply have not set the PROTO_PATH correctly.
If I look at your proto path in nodejs
```
PROTO_PATH = __dirname + './protos/example.proto'
```
vs your proto path in go
```
-I/usr/local/include -I. -I$GOPATH/src -I$GOPATH/src/github.com/grpc-ecosystem/grpc-gateway/third_party/googleapis
```
You might want to try adjusting your proto_path so that it can find gogoproto.proto

But this is just a guess and it might still depend on proto2 support.
